lib/Sema/TypeCheckDecl.cpp

```diff
@@ -3996,6 +3996,125 @@ static void finalizeAbstractStorageDecl(TypeChecker &TC,
   }
 }
 
+static void adjustFunctionTypeForOverride(Type &type) {
+  // Drop 'throws'.
+  // FIXME: Do we want to allow overriding a function returning a value
+  // with one returning Never?
+  auto fnType = type->castTo<AnyFunctionType>();
+  auto extInfo = fnType->getExtInfo();
+  extInfo = extInfo.withThrows(false);
+  if (fnType->getExtInfo() != extInfo)
+    type = fnType->withExtInfo(extInfo);
+}
+
+/// Drop the optionality of the result type of the given function type.
+static Type dropResultOptionality(Type type, unsigned uncurryLevel) {
```

Comment:
Structurally, this is almost identical to `TypeBase::replaceCovariantResultType()`. Should we generalize that with some kind of "here's a function object to transform the result type" thing?

Reply:
Sure, I will look into that separately!

Reply:
replaceCovariantResultType() is kind of weird because it completely ignores the result type. It's used for DynamicSelf on class methods, and I think it's kind of doing the wrong thing anyway -- in its current form it can't return a class method returning (Self, Self) or anything with DynamicSelf in a non-trivial but still covariant position. I wouldn't use it as a model for anything, it should instead go away.